examples/cis2-dynamic-nft/src/lib.rs

```diff
@@ -0,0 +1,1450 @@
+    ContractOnly,
+    /// Failed to invoke a contract.
+    InvokeContractError,
+    /// Failed to upgrade token because of max metadata counter is reached.
+    UpgradeTokenFailedError,
+}
+
+type ContractError = Cis2Error<CustomContractError>;
+
+type ContractResult<A> = Result<A, ContractError>;
+
+/// Mapping the logging errors to ContractError.
+impl From<LogError> for CustomContractError {
+    fn from(le: LogError) -> Self {
+        match le {
+            LogError::Full => Self::LogFull,
+            LogError::Malformed => Self::LogMalformed,
+        }
+    }
+}
+
+/// Mapping errors related to contract invocations to CustomContractError.
+impl<T> From<CallContractError<T>> for CustomContractError {
+    fn from(_cce: CallContractError<T>) -> Self { Self::InvokeContractError }
+}
+
+/// Mapping CustomContractError to ContractError
+impl From<CustomContractError> for ContractError {
+    fn from(c: CustomContractError) -> Self { Cis2Error::Custom(c) }
+}
+
+impl From<NewReceiveNameError> for CustomContractError {
+    fn from(_: NewReceiveNameError) -> Self { Self::InvalidContractName }
+}
+
+impl From<NewContractNameError> for CustomContractError {
+    fn from(_: NewContractNameError) -> Self { Self::InvalidContractName }
+}
+
+impl<S: HasStateApi> State<S> {
+    /// Construct a state with no tokens
+    fn empty(state_builder: &mut StateBuilder<S>) -> Self {
+        State {
+            state:        state_builder.new_map(),
+            tokens:       state_builder.new_map(),
+            implementors: state_builder.new_map(),
+        }
+    }
+
+    /// Mints an amount of tokens with a given address as the owner.
+    fn mint(
+        &mut self,
+        token_id: &ContractTokenId,
+        mint_param: &MintParam,
+        owner: &Address,
+        state_builder: &mut StateBuilder<S>,
+    ) {
+        self.tokens.insert(*token_id, TokenMetadataState {
+            token_metadata_current_state_counter: 0,
+            token_metadata_list:                  mint_param.metadata_url.clone(),
+        });
+        let mut owner_state =
+            self.state.entry(*owner).or_insert_with(|| AddressState::empty(state_builder));
+        let mut owner_balance = owner_state.balances.entry(*token_id).or_insert(0.into());
+        *owner_balance += mint_param.token_amount;
+    }
+
+    /// Picks the next MetadataUrl from the list when token is upgraded by
+    /// contract owner
+    fn upgrade_token(&mut self, token_id: &ContractTokenId) -> ContractResult<()> {
+        ensure!(self.contains_token(token_id), ContractError::InvalidTokenId);
+
+        let counter = self.get_token_state_medata_counter(token_id);
+        ensure!(
+            counter < self.tokens.get(token_id).unwrap().token_metadata_list.len(),
+            ContractError::Custom(CustomContractError::UpgradeTokenFailedError)
+        );
+
+        self.tokens.get_mut(token_id).unwrap().token_metadata_current_state_counter += 1;
```

Comment:
Suggestion: Since you first read the `counter` and then change/mutate the counter, this can be simplified by getting a mutable reference via the `entry`.

https://docs.rs/concordium-std/latest/concordium_std/struct.StateMap.html#method.entry
https://github.com/Concordium/concordium-rust-smart-contracts/blob/main/examples/cis3-nft-sponsored-txs/src/lib.rs#L802
```
self.tokens.entry(token_id)
```

Reply:
I can update the line with the following for incrementing, but does it really make a difference? 
`self.tokens.entry(*token_id).and_modify(|token_metadata_state|` {
            token_metadata_state.token_metadata_current_state_counter += 1
        });`

Reply:
This function uses several times `self.tokens.get(token_id)` and `self.tokens.get_mut(token_id)`. Using `self.token.entry(token_id)` at the top level of this function would allow us to read and mutate the value without having to repeat the code. 

Nonetheless, closing the thread since we can do minor optimizations in later iterations.